proposals/configurable_tooling.md

```diff
@@ -0,0 +1,111 @@
+  1. Lengthy setup instructions to create a DL project
+  1. Configurations and other setup requiring devops support
+  1. Exposure of the underlying cloud platform
+  1. An ad-hoc set of commands that do not compose well, may overlap or provide redundant or non-standard arguments.
+  1. Supporting commands that need to be installed on a data scientist's client machine (often with upgrade constraints).
+  1. Lack of command completion (ks) that does not integrate well with other commands or within a pipeline of commands.
+  1. Lack of an ability to integrate a data scientist customizations or integrate new CLI subcommands required by new kubeflow components.  
+
+## Goals
+- Provide a data scientist's CLI that:
+  1. Allows the data scientist to easily create a project or switch between projects.
+  1. Is able to discover available services such as dashboards, jobs, job histories, notebooks.
+  1. Is easily installed on the client machine.
+  1. Provides common kubeflow operations as customizable subcommands (train, serve, notebook)
+  1. Allows a data scientist or contributor to add or override subcommands using python
+  1. Allows subcommands to be added to only those granted particular privileges
+  1. Enables a subcommand registry   
+
+## Non-Goals
+The CLI subcommands suggested do not proscribe against defining additional subcommands in other proposals.
+
+## UI or API
+1. A CLI with an initial base set of subcommands:    
+   1. General    
+      1. `kf create [-h] [-d] [-r repo] [-b branch] [-i] <namespace>`
+      1. `kf delete [-h] <namespace>`
+      1. `kf remove [-h] <component>`
+      1. `kf run [-h] <component> [container] -- [command]`
+      1. `kf status [-h] [list|health|info][<component>]`
+      1. `kf use [-h] <namespace>`
+      1. `kf whoami [-h]`
+   1. Specific to Kubeflow components but not addressed per se within this proposal    
+      1. `kf notebook [-h] [ls|start|stop] <name>`
+      1. `kf serve [-h] <name> <yaml>`
+      1. `kf shell [-h] <component> [<container>]`
+      1. `kf tensorboard [-h] <name> <yaml>`
+      1. `kf train [-h] <name> <yaml>`
+1. A CLI that is extensible via a subcommand called `command`    
+  1. `kf command [-h] [add|remove|replace|upgrade|list|describe] <subcommand>`
+1. Subcommands call a Kubeless Function (written in python|bash|golang) which:    
+   1. Runs in a container that includes requisite tooling and dependencies like ks, kubeclt, kubeless, python2.7|python3.6
+
+![CLI calls a serverless Function](diagrams/CLI2Function.png)
+## Design
```

Comment:
Besides CLI I'd love to see some API so people at some point could write custom software driving kubeflow - like jupyter plugins, CI systems etc. That will also be useful for automatic hyperparameter tuning work (just call api "run" with different hyperparameters).

Reply:
kubeless has some pretty cool features including a CRD operator and events - take a look https://github.com/kubeless/functions/tree/master/incubator. Also there is an opportunity to rework the function operator to address other needs (and it more like a kubeflow operator) see https://github.com/kubeless/kubeless/tree/master/pkg/controller